### PR TITLE
New mixed phase property functions for Helmholtz EoS.

### PR DESCRIPTION
**Pull request by eslickj**
## Summary/Motivation:

This changes the mixed phase properties for Helmholtz EoS to directly call external functions for properties as a function of (h, P), (s, p), or (u, p) depending on the state varaible selection.  The goal of this is to prevent calls to functions for properties of phases that don't exist.  While I'm not aware of any issues calling functions for properties in phases that don't exist, this should be an improvement.  It may make things solver faster by reducing the number of external function calls, and potentially it could improve robustness.

The CI tests will fail until I get the new binaries up with the new functions calls.  I did test this myself, and everything seems to work.  It is probably okay to review despite the tests not running.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py [hunk 1]
```diff
@@ -154,12 +154,165 @@ class AmountBasis(enum.Enum):
...
+    "w_func": {  # speed of sound
         "fname": "w",
         "units": pyo.units.m / pyo.units.s,
         "arg_units": [dimensionless, dimensionless, dimensionless],
         "doc": "w(comp, delta, tau)",
     },
+    "v_func": {  # specific volume
+        "fname": "v",
+        "units": pyo.units.m**3 / pyo.units.kg,
+        "arg_units": [dimensionless, pyo.units.kg / pyo.units.kg, pyo.units.kPa],
```
Comment (dallan-keylogic) [inline idaes/models/properties/general_helmholtz/helmholtz_functions.py:166]: Delta and Tau are some normalized (and possibly inverted?) versions of density and temperature, respectively, right? Assuming "arg_units" means, "units with which the function is called", we've got an issue here.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py [hunk 2]
```diff
@@ -154,12 +154,165 @@ class AmountBasis(enum.Enum):
...
+    "h_sp_func": {  # enthalpy
+        "fname": "h_sp",
+        "units": pyo.units.kJ / pyo.units.kg,
+        "arg_units": [dimensionless, pyo.units.kJ / pyo.units.kg/ pyo.units.K, pyo.units.kPa],
+        "doc": "h(comp, entropy, pressure)",
+    },
+    "g_sp_func": {  # Gibbs free energy
+        "fname": "g_sp",
+        "units": pyo.units.kJ / pyo.units.kg,
+        "arg_units": [dimensionless, pyo.units.kJ / pyo.units.kg/ pyo.units.K, pyo.units.kPa],
```
Comment (dallan-keylogic) [inline idaes/models/properties/general_helmholtz/helmholtz_functions.py:234]: These units also look wrong

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py [hunk 3]
```diff
@@ -481,6 +634,31 @@ def _sv_str(**kwargs):
...
+                if self.amount_basis == AmountBasis.MOLE:
+                    s = kwargs["s"] * self.param.uc["J/mol/K to kJ/kg/K"]
+                else:
+                    s = kwargs["s"] * self.param.uc["J/kg/K to kJ/kg/K"]
+                return StateVars.PS, blk, c, s, p
+            if "u" in kwargs and kwargs["u"] is not None:
+                if self.amount_basis == AmountBasis.MOLE:
+                    u = kwargs["u"] * self.param.uc["J/mol to kJ/kg"]
+                else:
+                    u = kwargs["u"] * self.param.uc["J/kg to kJ/kg"]
```
Comment (dallan-keylogic) [inline idaes/models/properties/general_helmholtz/helmholtz_functions.py:658]: All these are indented a level beyond what they should, I think

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py [hunk 4]
```diff
@@ -481,6 +634,31 @@ def _sv_str(**kwargs):
     def add_funcs(self, names=None):
         add_helmholtz_external_functions(self.blk, names=names)
 
+    def _state_vars(self, **kwargs):
```
Comment (dallan-keylogic) [inline idaes/models/properties/general_helmholtz/helmholtz_functions.py:637]: Could you add a docstring describing what's supposed to be returned here? It's hard to follow the code further on without knowing what "c", "u", etc. are.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 1 above)
Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:166]: Yeah, those units are wrong.  I don't actually use that function anywhere yet so it did cause a problem.  Will fix.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 2 above)
Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:234]: I think these are right.  The args art component (dimensionless), entropy (kJ/kg/K), and pressure kPa.  The Gibbs free energy is kJ/kg.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 4 above)
Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:637]: Yes

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 2 above)
Comment (dallan-keylogic) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:234]: The doc directly below says `"g(comp, delta, tau)"`. Maybe the variables are what are wrong?

Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:234]: Oh yeah the doc string is wrong.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 3 above)
Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:658]: I think it's right they are in the "if p is an args."  I will double check everything in the function again though.

Comment (dallan-keylogic) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:658]: So all of those conversions should happen only if pressure is a state variable and also converted?

Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:658]: I used to have (s, T), but as it is pressure is always a state variable. Could change in the future if we find a reason.

**Review by andrewlee94 — APPROVED** [review]

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py [hunk 5]
```diff
@@ -1438,14 +1768,17 @@ def build(self):
             "J/mol to J/kg": 1.0 / self.mw,
             "kJ/kg to J/mol": (pyo.units.J * 1000 / pyo.units.kJ) * self.mw,
             "J/mol/K to kJ/kg/K": (pyo.units.kJ / 1000 / pyo.units.J) / self.mw,
-            "J/mol/K to J/kg/K": 1.0 / self.mw,
+            "J/mol/K to J/kg/K": 1 / self.mw,
```
Comment (dallan-keylogic) [inline idaes/models/properties/general_helmholtz/helmholtz_functions.py:1771]: I seem to recall @andrewlee94 replacing a bunch of integer zeros or ones with floating point zeros or ones somewhere in core. Is this a place where that distinction matters?

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py [hunk 6]
```diff
@@ -563,12 +782,22 @@ def basic_calculations(
 
     def s(self, **kwargs):
         """Mixed phase entropy"""
-        blk, delta_liq, delta_vap, tau, x, c = self.basic_calculations(**kwargs)
-        self.add_funcs(names=["s_func"])
-        s = (
-            blk.s_func(c, delta_liq, tau, _data_dir) * (1 - x)
-            + blk.s_func(c, delta_vap, tau, _data_dir) * x
-        )
+        sv, blk, c, u, p = self._state_vars(**kwargs)
```
Comment (dallan-keylogic) [inline idaes/models/properties/general_helmholtz/helmholtz_functions.py:785]: It bugs me that the first state variable is always assigned to `u`, even though it could be any one of internal energy, entropy, or enthalpy. I think confusion around this point was why I gave up reviewing this PR the first time. It's doesn't affect the code's functionality, but it's bad for the maintainability to use confusing naming conventions.

**Review by dallan-keylogic — APPROVED** [review]

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 5 above)
Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:1771]: I'm pretty sure it doesn't matter here since the division will make it a float anyway.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py [hunk 7]
```diff
@@ -770,12 +1067,34 @@ def x(self, **kwargs):
...
+        if sv == StateVars.PH:
+            self.add_funcs(names=["tau_func"])
+            tau = blk.tau_func(c, u, p, _data_dir)
+        elif sv == StateVars.PU:
+            self.add_funcs(names=["tauu_func"])
+            tau = blk.tauu_func(c, u, p, _data_dir)
+        elif sv == StateVars.PS:
+            self.add_funcs(names=["taus_func"])
+            tau = blk.taus_func(c, u, p, _data_dir)
```
[the remark this answers is not in the document]
Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:1079]: I in the beginning there was only tau as a function of h and p.  I was just afraid to change it.  Maybe I'll do another PR to clean some things up.  Too many moving parts confuse me.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 5 above)
Comment (andrewlee94) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:1771]: It also doesn't really matter unless you want to look at the expression body itself.

File: idaes/models/properties/general_helmholtz/helmholtz_functions.py (hunk 6 above)
Comment (eslickj) [reply idaes/models/properties/general_helmholtz/helmholtz_functions.py:785]: Noted.  I'm planing to follow this up soon that just uses direct property calls for everything without going through delta and tau.  I'll change it then.
